logic/path/src/selectors/dfs.rs

```diff
@@ -74,86 +94,104 @@ impl EdgeWeighting<U256> for RandomizedEdgeWeighting {
-where
-    CW: EdgeWeighting<U256>,
-{
-    /// Determines whether a node is considered "interesting".
+impl<CW: EdgeWeighting<U256>> DfsPathSelector<CW> {
+    /// Creates new path selector with the given [config](DfsPathSelectorConfig).
+    pub fn new(cfg: DfsPathSelectorConfig) -> Self {
+        Self { cfg, _cw: PhantomData }
+    }
+
+    /// Determines whether a `next_hop` node is considered "interesting".
     ///
     /// To achieve privacy, we need at least one honest node along
     /// the path. Each additional node decreases the probability of
     /// having only malicious nodes, so we can sort out many nodes.
     /// Nodes that have shown to be unreliable are of no use, so
     /// drop them.
-    fn filter_channel(
+    ///
+    /// ## Arguments
+    /// * `next_hop`: node in the channel graph that we're trying to add to the path
+    /// * `edge`: the information about the corresponding graph edge ([`ChannelEntry`] and `score`).
+    /// * `initial_source`: the initial node on the path
+    /// * `final_destination`: the desired destination node (will not be part of the path)
+    /// * `current_path`: currently selected relayers
+    fn is_next_hop_usable(
         &self,
-        channel: &ChannelEdge,
-        source: &Address,
-        destination: &Address,
+        next_hop: &Node,
+        edge: &ChannelEdge,
+        initial_source: &Address,
+        final_destination: &Address,
         current_path: &[Address],
     ) -> bool {
-        if source.eq(&channel.channel.destination) {
-            // looping back to self does not give any privacy
+        debug_assert_eq!(next_hop.address, edge.channel.destination);
+
+        // Looping back to self does not give any privacy
+        if next_hop.address.eq(initial_source) {
+            trace!(%next_hop, "source loopback not allowed");
+            return false;
+        }
+
+        // We cannot use `final_destination` as last intermediate hop as
+        // this would be a loopback that does not give any privacy
+        if next_hop.address.eq(final_destination) {
+            trace!(%next_hop, "destination loopback not allowed");
             return false;
         }
 
-        if destination.eq(&channel.channel.destination) {
-            // We cannot use destination as last intermediate hop as
-            // this would be a loopback that does not give any privacy
+        // Only use nodes that have shown to be somewhat reliable
+        if next_hop.quality < self.cfg.quality_threshold {
+            trace!(%next_hop, "node quality threshold not satisfied");
             return false;
         }
 
-        // Check node reliability, new nodes are NOT considered reliable yet
-        if channel.quality.unwrap_or(0.0f64) < self.quality_threshold {
-            // Only use nodes that have shown to be somewhat reliable
+        // Edges which have score below the threshold won't be considered
+        if edge.score.unwrap_or(1.0) < self.cfg.score_threshold {
```

Comment:
Why is the default `1.0` here?

Reply:
We're not updating the score in the graph right now, so the threshold is 0 and default is 1, so that it has no effect.

Reply:
I changed it slightly, so it has more clarity and does not check the criteria when no `score` is set.